### PR TITLE
Add extended metrics ( from logs )

### PR DESCRIPTION
**Pull request by cpaillet**
- We cannot have metrics by OriginIP on the default metrics, we need
  to fetch logs ( last 5 minutes ) and create metrics from those.

  We have now metric for each Cloudflare-DC, to each Origin-IP
  We have the OriginResponseTime, and a count.

### CODE REVIEW

File: cloudflare_exporter/collector.py
Comment (pierrecdn) [inline cloudflare_exporter/collector.py]: Nothing else available? _unix_ or _rfc3339_ to avoid nanos2s then and map this easily to a datetime?

**Review by pierrecdn — CHANGES_REQUESTED** [review]
Sounds to me that we're doing a log parser or a log2metrics solution here, is it scalable?
Won't it hurt a lot other/existing metrics processing?

File: requirements.txt
Comment (pierrecdn) [inline requirements.txt]: Numpy is a 20MB  library.. Do we have a real use-case for importing this?

File: cloudflare_exporter/collector.py
Comment (cpaillet) [reply cloudflare_exporter/collector.py]: > Sounds to me that we're doing a log parser or a log2metrics solution here, is it scalable?
Do you know good lib for that ?
We just want some sampling here, it's enough. We limit the number of line read, no scalling problem.
> Won't it hurt a lot other/existing metrics processing?
I don't understand your question

Comment (cpaillet) [reply cloudflare_exporter/collector.py]: > 
> 
> Nothing else available? _unix_ or _rfc3339_ to avoid nanos2s then and map this easily to a datetime?

the OriginResponseTime is allways in nano... ( even If I set timestamps to _unix_ or _rfc3339_ )
then I remove this line

File: requirements.txt
Comment (cpaillet) [reply requirements.txt]: it's use from percentille. do you know a smaller lib for that ?

File: cloudflare_exporter/collector.py
Comment (pierrecdn) [reply cloudflare_exporter/collector.py]: * _rfc3339_: my point was this one:
```
datetime.datetime.fromisoformat('2017-05-08T17:30:34.009056+05:30')
>>> datetime.datetime(2017, 5, 8, 17, 30, 34, 9056, tzinfo=datetime.timezone(datetime.timedelta(seconds=19800)))
```
We could avoid writing our own `nano2s` conversion method.

* About scalability, I asked about the performance of this operation of log parsing. What will happen when the API will return thousands/millions of logs and MBs of output? What will happen if this happens when the scraping from prometheus is in progress? Does this should be done in a background task? etc etc.

File: requirements.txt
Comment (pierrecdn) [reply requirements.txt]: The stdlib? https://docs.python.org/3/library/statistics.html

Comment (cpaillet) [reply requirements.txt]: quantiles ? -> new in 3.8

File: cloudflare_exporter/collector.py
Comment (cpaillet) [reply cloudflare_exporter/collector.py]: There is a limit -> 'count': LOGS_EXTENDED_METRIC_COUNT,  
We just need a sample, we don't need all the log to make those stats

Comment (cpaillet) [reply cloudflare_exporter/collector.py]: I receive from clouflare
OriginResponseTime=136000000  #136ms
do you have an easier method to change it to second ?

Comment (pierrecdn) [reply cloudflare_exporter/collector.py]: OK, my understanding from https://developers.cloudflare.com/logs/logpull-api/requesting-logs/, was that they unified timestamps by using date everywhere. Got it now, thanks!

File: requirements.txt
Comment (pierrecdn) [reply requirements.txt]: OK, fair point. Still I'd require an advice from someone on #python for ex, it's really worth avoiding 20MB libs just for the purpose of doing quantiles..

**Review by pierrecdn — CHANGES_REQUESTED** [review]
Got it with your explanations!
Still the numpy thing for which we should find a better solution IMO.

**Comment by cpaillet** [comment]
> 
> 
> Got it with your explanations!
> Still the numpy thing for which we should find a better solution IMO.

I add a TODO, is it ok for you ?

**Review by dclaisse — COMMENTED** [review]
LGTM. Waiting for Pierre's approval, otherwise ok for me.

**Review by pierrecdn — APPROVED** [review]

**Pull request merged**
